# Benign vs Malignant

![alt text](https://www.verywellhealth.com/thmb/FdgmCw-TjROLOXO6bsUtzxPpdYc=/700x0/filters:no_upscale():max_bytes(150000):strip_icc()/514240-article-img-malignant-vs-benign-tumor2111891f-54cc-47aa-8967-4cd5411fdb2f-5a2848f122fa3a0037c544be.png)

## Load Dataset

In [1]:
import numpy as np
import sklearn.datasets 

  Check the url : 
  https://www.kaggle.com/uciml/breast-cancer-wisconsin-data 
        for the official dataset


In [2]:
dataset = sklearn.datasets.load_breast_cancer()

In [3]:
dataset

{'DESCR': '.. _breast_cancer_dataset:\n\nBreast cancer wisconsin (diagnostic) dataset\n--------------------------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry \n        - fractal dimension ("coastline approximation" - 1)\n\n        The mean, standard error, and "worst" or largest (mean of the three\n        largest values) of these features were computed for each image,\n        resulting in 30 features.  For

In [4]:
X = dataset.data
y = dataset.target

In [5]:
X

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [6]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [7]:
print(X.shape, y.shape)

(569, 30) (569,)


So we have **569** observations i.e **569 rows**      
      with **30** features  i.e **30 cols**

## Generate a DataFrame

In [8]:
import pandas as pd

In [9]:
data = pd.DataFrame(dataset.data, columns = dataset.feature_names)

In [10]:
data['class'] = dataset.target 

In [11]:
data.head()

mean radius  mean texture  ...  worst fractal dimension  class
0        17.99         10.38  ...                  0.11890      0
1        20.57         17.77  ...                  0.08902      0
2        19.69         21.25  ...                  0.08758      0
3        11.42         20.38  ...                  0.17300      0
4        20.29         14.34  ...                  0.07678      0

[5 rows x 31 columns]

In [12]:
data.head().T

0            1  ...           3            4
mean radius                17.990000    20.570000  ...   11.420000    20.290000
mean texture               10.380000    17.770000  ...   20.380000    14.340000
mean perimeter            122.800000   132.900000  ...   77.580000   135.100000
mean area                1001.000000  1326.000000  ...  386.100000  1297.000000
mean smoothness             0.118400     0.084740  ...    0.142500     0.100300
mean compactness            0.277600     0.078640  ...    0.283900     0.132800
mean concavity              0.300100     0.086900  ...    0.241400     0.198000
mean concave points         0.147100     0.070170  ...    0.105200     0.104300
mean symmetry               0.241900     0.181200  ...    0.259700     0.180900
mean fractal dimension      0.078710     0.056670  ...    0.097440     0.058830
radius error                1.095000     0.543500  ...    0.495600     0.757200
texture error               0.905300     0.733900  ...    1.156000     0.781300
perimeter error             8.589000     3.398000  ...    3.445000     5.438000
area error                153.400000    74.080000  ...   27.230000    94.440000
smoothness error            0.006399     0.005225  ...    0.009110     0.011490
compactness error           0.049040     0.013080  ...    0.074580     0.024610
concavity error             0.053730     0.018600  ...    0.056610     0.056880
concave points error        0.015870     0.013400  ...    0.018670     0.018850
symmetry error              0.030030     0.013890  ...    0.059630     0.017560
fractal dimension error     0.006193     0.003532  ...    0.009208     0.005115
worst radius               25.380000    24.990000  ...   14.910000    22.540000
worst texture              17.330000    23.410000  ...   26.500000    16.670000
worst perimeter           184.600000   158.800000  ...   98.870000   152.200000
worst area               2019.000000  1956.000000  ...  567.700000  1575.000000
worst smoothness            0.162200     0.123800  ...    0.209800     0.137400
worst compactness           0.665600     0.186600  ...    0.866300     0.205000
worst concavity             0.711900     0.241600  ...    0.686900     0.400000
worst concave points        0.265400     0.186000  ...    0.257500     0.162500
worst symmetry              0.460100     0.275000  ...    0.663800     0.236400
worst fractal dimension     0.118900     0.089020  ...    0.173000     0.076780
class                       0.000000     0.000000  ...    0.000000     0.000000

[31 rows x 5 columns]

In [13]:
data.describe()

mean radius  mean texture  ...  worst fractal dimension       class
count   569.000000    569.000000  ...               569.000000  569.000000
mean     14.127292     19.289649  ...                 0.083946    0.627417
std       3.524049      4.301036  ...                 0.018061    0.483918
min       6.981000      9.710000  ...                 0.055040    0.000000
25%      11.700000     16.170000  ...                 0.071460    0.000000
50%      13.370000     18.840000  ...                 0.080040    1.000000
75%      15.780000     21.800000  ...                 0.092080    1.000000
max      28.110000     39.280000  ...                 0.207500    1.000000

[8 rows x 31 columns]

## EDA - Explaratory Data Analysis

In [14]:
data['class'].value_counts()

1    357
0    212
Name: class, dtype: int64

In [15]:
dataset.target_names

array(['malignant', 'benign'], dtype='<U9')

We have **357 (class 0 - array[0])** malignant cases
and **212 (class 1 - array[1])** beningn cases

In [16]:
data.groupby('class').mean()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
class                             ...                                         
0        17.462830     21.604906  ...        0.323468                 0.091530
1        12.146524     17.914762  ...        0.270246                 0.079442

[2 rows x 30 columns]

## Train - Test Split

We split the data to check whether our model has generalised to data other than the train data

In [17]:
from sklearn.model_selection import train_test_split

We check this so that our datset is uniformly distributed

Stratify parameter will preserve the proportion of target as in original dataset, in the train and test datasets as well.

So if your original dataset df has target/label as [0,1,2] in the ratio say, 40:30:30. That is, for every 100 datasets, you can find 40, 30 and 30 observations of target 0,1 and 2 respectively.

Now when you split this original using the train_test_split(x,y,test_size=0.1,stratify=y), the methods returns train and test datasets in the ratio of 90:10. Now in each of these datasets, the target/label data proportion is preserved as 40:30:30 for the classes [0,1,2].

Often, we want to preserve the dataset proportions for better prediction and reproduceability of results  

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, stratify =y)

In [19]:
print("X_train shape : ", X_train.shape)
print("y_train shape : ", y_train.shape)
print("X_test shape  : ", X_test.shape)
print("y_test shape  : ", y_test.shape)

X_train shape :  (512, 30)
y_train shape :  (512,)
X_test shape  :  (57, 30)
y_test shape  :  (57,)


In [20]:
57/569

0.10017574692442882

So we got a 90/10 % split

In [21]:
print(y.shape, y_train.shape, y_test.shape)

(569,) (512,) (57,)


In [22]:
print(y.mean(), y_train.mean(), y_test.mean())

0.6274165202108963 0.626953125 0.631578947368421


## Build the KNN Classifier

![alt text](https://www.kdnuggets.com/wp-content/uploads/knn2.jpg)


![alt text](http://res.cloudinary.com/dyd911kmh/image/upload/f_auto,q_auto:best/v1531424125/KNN_final1_ibdm8a.png)

![alt text](https://www.kdnuggets.com/wp-content/uploads/knn1.png)

In [23]:
from sklearn.neighbors import KNeighborsClassifier

In [34]:
def build(n=3):

  model = KNeighborsClassifier(n_neighbors=n )
  return model

## Train the model

In [30]:
from sklearn.metrics import accuracy_score

## Selecting the best K for KNN

In [35]:
score = []
for i in range(1,10):
  model = build(i)
  model.fit(X_train , y_train)
  print(" Model with k = ",i,
        "\n\tTrain_accuracy : ", accuracy_score(y_train,model.predict(X_train)),
        "\n\tTest accuracy  : ", accuracy_score(y_test, model.predict(X_test)))
  score.append((accuracy_score(y_test, model.predict(X_test))))

 Model with k =  1 
	Train_accuracy :  1.0 
	Test accuracy  :  0.9649122807017544
 Model with k =  2 
	Train_accuracy :  0.96875 
	Test accuracy  :  0.9824561403508771
 Model with k =  3 
	Train_accuracy :  0.951171875 
	Test accuracy  :  0.9649122807017544
 Model with k =  4 
	Train_accuracy :  0.94140625 
	Test accuracy  :  0.9824561403508771
 Model with k =  5 
	Train_accuracy :  0.939453125 
	Test accuracy  :  0.9649122807017544
 Model with k =  6 
	Train_accuracy :  0.939453125 
	Test accuracy  :  0.9824561403508771
 Model with k =  7 
	Train_accuracy :  0.939453125 
	Test accuracy  :  0.9824561403508771
 Model with k =  8 
	Train_accuracy :  0.935546875 
	Test accuracy  :  0.9824561403508771
 Model with k =  9 
	Train_accuracy :  0.939453125 
	Test accuracy  :  0.9824561403508771


In [40]:
best = np.asarray(score)

In [41]:
best.max()

0.9824561403508771

In [44]:
best_k = best.argmax()+1
best_k

2

## Build using the best K value

In [46]:
model = build(best_k)
model.fit(X_train, y_train)

predict_on_train = model.predict(X_train)
print("Accuracy on train :",accuracy_score(y_train,predict_on_train))
predict_on_test = model.predict(X_test)
print("Accuracy on val :",accuracy_score(y_test,predict_on_test))


Accuracy on train : 0.96875
Accuracy on val : 0.9824561403508771


## Check on out of dataset i.e actual test set

The observation is taken from kaggle for a malignant patient 0

In [47]:
input_data = (17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,
              0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,
              0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,
              0.6656,0.7119,0.2654,0.4601,0.1189)

In [48]:
len(input_data)

30

In [49]:
array = np.asarray(input_data)

In [50]:
array.shape

(30,)

Reshape the array as it's a single observation whereas the model is trained to make prediction on lot of observations

In [51]:
reshaped = array.reshape(1,-1)

In [52]:
reshaped.shape

(1, 30)

In [53]:
out = model.predict(reshaped)

In [54]:
if out[0] ==0:
  print("Malignant, Patient is Fine")
else:
  print("Beningn, Fatal Danger")

Malignant, Patient is Fine


Check for a beningn case

In [55]:
input_2 = (9.504,12.44,60.34,273.9,0.1024,0.06492,0.02956,0.02076,0.1815,0.06905,0.2773,0.9768,1.909,15.7,0.009606,0.01432,0.01985,0.01421,0.02027,0.002968,10.23,15.66,65.13,314.9,0.1324,0.1148,0.08867,0.06227,0.245,0.07773)
array_2 = np.asarray(input_2)
reshaped_2 = array_2.reshape(1,-1)

out_2 = model.predict(reshaped_2)

if out_2[0] ==0:
  print("Malignant, Patient is Fine")
else:
  print("Beningn, Fatal Danger")


Beningn, Fatal Danger
